### PR TITLE
new(userspace/libsinsp): allow to specify a list of cri socket paths to be tried in fallback manner.

### PR DESCRIPTION
**Pull request by FedeDP**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-ebpf

> /area libscap

/area libsinsp

> /area tests

> /area proposals

**What this PR does / why we need it**:

This allows to pass multiple "--cri" option in Falco and be sure that the first socket that works will be used.
A PR in Falco is needed to manage multiple "--cri" options though.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
new: support setting multiple cri socket paths to be tried one after the other until a working one is found
```

### CODE REVIEW

File: userspace/libsinsp/container_engine/cri.cpp [hunk 1]
```diff
@@ -207,20 +207,35 @@ bool cri_async_source::lookup_sync(const libsinsp::cgroup_limits::cgroup_limits_
...
+		struct stat s = {};
+		if(stat(cri_path.c_str(), &s) != 0 || (s.st_mode & S_IFMT) != S_IFSOCK)
+		{
+			continue;
+		}
+
+		m_cri = std::unique_ptr<libsinsp::cri::cri_interface>(new libsinsp::cri::cri_interface(cri_path));
+		if(!m_cri->is_ok())
+		{
+			m_cri.reset(nullptr);
```
Comment (jasondellaluce) [inline userspace/libsinsp/container_engine/cri.cpp:233]: Maybe I'm not seeing the bigger picture, but don't we risk adding the same CRI socket path twice in this loop?

**Review by jasondellaluce — COMMENTED** [review]
LGTM! Left only one comment.

File: userspace/libsinsp/container_engine/cri.cpp (hunk 1 above)
Comment (FedeDP) [reply userspace/libsinsp/container_engine/cri.cpp:233]: Mmh i don't get it: as soon as a working socket is found, we use it and break out of the loop!
Am i missing anything obvious?

Comment (jasondellaluce) [reply userspace/libsinsp/container_engine/cri.cpp:233]: Yeah maybe a comment to make this explicit would be good. Either way, I agree this is definitely something we can avoid supporting for now.

Comment (leogr) [reply userspace/libsinsp/container_engine/cri.cpp:233]: Yeah, the rationale is that having multiple container runtimes on the same host is a sporadic case, so we wouldn't make things complex to support that. On the other hand, specifying the `--cri` multiple times (and using only the first match) will solve the "same config, multiple hosts" use case.

**Review by leogr — APPROVED** [review]
LGTM!

Maybe we can add some more comments (as suggested [here](https://github.com/falcosecurity/libs/pull/202/files#r805874653)). Anyway, not a blocker for me.

/approve

File: userspace/libsinsp/container_engine/cri.cpp (hunk 1 above)
Comment (FedeDP) [reply userspace/libsinsp/container_engine/cri.cpp:233]: Should i document it somewhere in this PR?

**Review by leogr — DISMISSED** [review]
/approve

File: userspace/libsinsp/cri.h [hunk 2]
```diff
@@ -37,9 +37,8 @@ namespace libsinsp {
 namespace cri {
 
 // these shouldn't be globals but we still need references to *the* CRI runtime
-extern std::string s_cri_unix_socket_path;
+extern std::vector<std::string> s_cri_unix_socket_paths;
 extern int64_t s_cri_timeout;
-extern sinsp_container_type s_cri_runtime_type;
```
Comment (gnosek) [inline userspace/libsinsp/cri.h:42]: Can we *not* drop these two? `s_cri_unix_socket_path` should point to the one socket path we ended up using.

Comment (FedeDP) [reply userspace/libsinsp/cri.h:42]: @gnosek it should be ok now.

**Review by gnosek — APPROVED** [review]
LGTM, thanks!

**Review by leogr — APPROVED** [review]
/approve

**Pull request merged**